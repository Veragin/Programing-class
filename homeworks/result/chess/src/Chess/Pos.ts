export const letters = ["A", "B", "C", "D", "E", "F", "G", "H"] as TY[];

type TPos = { x: TX; y: TY };
type TIndex = { i: number; j: number };

export class Pos {
    private _i: number;
    private _j: number;

    constructor(input: TPos | TIndex) {
        const inp = input as any;
        if (inp.x !== undefined) {
            this._i = inp.x - 1;
            this._j = letters.indexOf(inp.y);
        } else {
            this._i = inp.i;
            this._j = inp.j;
        }
    }

    get x() {
        return (this._i + 1) as TX;
    }
    get y() {
        return letters[this._j];
    }
    get i() {
        return this._i;
    }
    get j() {
        return this._j;
    }

    isEqual(pos: Pos) {
        return this.i === pos.i && this.j === pos.j;
    }

    export() {
        return [this._i, this._j];
    }
}
