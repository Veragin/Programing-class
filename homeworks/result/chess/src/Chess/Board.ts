import { BOARD_SIZE } from "./Const";
import { ChessPiece } from "./Pieces/ChessPiece";
import { Pos } from "./Pos";

export class Board {
    board: TBoard = [];

    constructor() {
        for (let i = 0; i < BOARD_SIZE; i++) {
            this.board.push([]);
            for (let j = 0; j < BOARD_SIZE; j++) {
                this.board[i].push(null);
            }
        }
    }

    set(piece: ChessPiece, pos: Pos) {
        try {
            const { i, j } = piece.findMyPosInBoard(this.board);
            this.board[i][j] = null;
        } catch {}

        this.board[pos.i][pos.j] = piece;
    }
}
