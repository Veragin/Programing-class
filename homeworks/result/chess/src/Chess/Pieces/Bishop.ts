import { Pos } from "../Pos";
import { isOnBoard } from "../utils";
import { ChessPiece } from "./ChessPiece";

const signs = [
    { i: 1, j: 1 },
    { i: 1, j: -1 },
    { i: -1, j: -1 },
    { i: -1, j: 1 },
];

export class Bishop extends ChessPiece {
    move(board: TBoard): Pos[] {
        const res: Pos[] = [];
        const { i, j } = this.findMyPosInBoard(board);

        for (let sign of signs) {
            let distance = 1;
            while (true) {
                const testI = i + sign.i * distance;
                const testJ = j + sign.j * distance;

                if (!isOnBoard(testI) || !isOnBoard(testJ)) {
                    break;
                }

                const pieceOnTile = board[testI][testJ];
                if (pieceOnTile === null) {
                    res.push(new Pos({ i: testI, j: testJ }));
                    distance++;
                    continue;
                }

                if (pieceOnTile.player !== this.player) {
                    res.push(new Pos({ i: testI, j: testJ }));
                }

                break;
            }
        }

        return res;
    }
}
