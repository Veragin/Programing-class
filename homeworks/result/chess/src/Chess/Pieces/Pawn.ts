import { Pos } from "../Pos";
import { ChessPiece } from "./ChessPiece";

export class Pawn extends ChessPiece {
    readonly type: TPieceType = "pawn";

    move(board: TBoard) {
        const res: Pos[] = [];
        const { i, j } = this.findMyPosInBoard(board);

        const newI = this.player === "black" ? i - 1 : i + 1;

        if (board[newI][j] === null) {
            res.push(new Pos({ i: newI, j }));
        }

        let p = board[newI][j - 1];
        if (p instanceof ChessPiece && p.player !== this.player) {
            res.push(new Pos({ i: newI, j: j - 1 }));
        }

        p = board[newI][j + 1];
        if (p instanceof ChessPiece && p.player !== this.player) {
            res.push(new Pos({ i: newI, j: j + 1 }));
        }

        return res;
    }
}
