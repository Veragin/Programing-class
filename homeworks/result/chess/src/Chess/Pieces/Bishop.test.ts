import { Board } from "../Board";
import { Pos } from "../Pos";
import { Bishop } from "./Bishop";
import { Pawn } from "./Pawn";

const createData = () => {
    const board = new Board();
    const bishop = new Bishop("white");
    const pos = new Pos({ x: 2, y: "B" });

    board.set(bishop, pos);

    return {
        board: board,
        bishop,
        pos,
    };
};

describe("Bishop", () => {
    it("can move", () => {
        const { board, bishop, pos } = createData();

        board.set(new Pawn("black"), new Pos({ i: pos.i + 2, j: pos.j + 2 }));

        const res = bishop.move(board.board);
        expect(res.length).toBe(5);

        const exp = res.map((pos) => pos.export());
        expect(exp).toMatchSnapshot();
    });

    it("cannt move if is blocked by figure", () => {
        const { board, bishop } = createData();
        board.set(bishop, new Pos({ x: 1, y: "H" }));

        board.set(new Pawn("white"), new Pos({ x: 6, y: "C" }));

        const res = bishop.move(board.board);
        expect(res.length).toBe(4);
    });
});

export {};
