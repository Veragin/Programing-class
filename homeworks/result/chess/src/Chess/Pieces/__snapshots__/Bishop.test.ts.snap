// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bishop can move 1`] = `
Array [
  Array [
    2,
    2,
  ],
  Array [
    3,
    3,
  ],
  Array [
    2,
    0,
  ],
  Array [
    0,
    0,
  ],
  Array [
    0,
    2,
  ],
]
`;
